pennylane/devices/device_api.py

```diff
@@ -0,0 +1,498 @@
+# Copyright 2018-2022 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""This module contains the Abstract Base Class for the next generation of devices.
+
+"""
+
+import abc
+
+from numbers import Number
+from typing import Callable, Union, Sequence, Tuple
+
+from pennylane.tape import QuantumTape
+from pennylane import Tracker
+
+from .execution_config import ExecutionConfig, DefaultExecutionConfig
+
+QuantumTapeBatch = Sequence[QuantumTape]
+QuantumTape_or_Batch = Union[QuantumTape, QuantumTapeBatch]
+
+Config_or_Batch = Union[ExecutionConfig, Sequence[ExecutionConfig]]
+
+# pylint: disable=unused-argument, no-self-use
+class DeviceDriver(abc.ABC):
```

Comment:
I think the main abstraction here is that PennyLane talks to a device (through a device driver) rather than holds/creates a device. All communication is mediated through the driver, which controls the underlying hardware/simulator.

This parlance is more in line with software-engineering practices, and is standard place for such behaviour (common for [all classical hardware integrations](https://www.oreilly.com/library/view/linux-device-drivers/0596005903/ch01.html), taking [CUDA devices](https://docs.nvidia.com/cuda/cuda-driver-api/index.html) as an example). I think this is one of those "better in the long run" situations, provided it is aptly described up front. Though, just my 2c.

Reply:
My 2 cents: I think we should double down on DeviceDriver exactly _because_ of these docs and user-facing things we have. I can't speak to the best/cleanest way to introduce it to users, but every time our code and docs say `device`, they're lying to the user and creating more misunderstanding. Driver is a more accurate term, and because this new API is about doing things the right way, we should do that.

Reply:
> because this new API is about doing things the right way
I'm starting to see more and more people motivating the new device work in these terms. It's certainly not our sole goal. Something we should aim to improve as we do this, for sure, but not the intrinsic motivation for all the effort we're likely to put in

Reply:
One decision I'd actually be OK with is *calling* it `QuantumDevice`, but *defining* it as a driver that controls one or more backends, where a backend can either be a QPU or virtual (simulator).

Reply:
thanks y'all for the feedback, it's something we discuss quite a bit so I definitely appreciate it. I'll admit "the Right Way" is a dangerous phrase to use. What I was really getting at is, when working with things like the old device interface, we find ourselves often having problems with some detail or another (in other words, "this detail gives me trouble"). There might not be a Right Way (is there ever?), but I think some of this project could be framed as "not doing things in the way that gives us trouble today".

On this specific topic: we internally struggle with the word "device" because we then talk about physical devices/simulators and go, "sooo that's also a 'Device'?" and then confuse ourselves. Hence, the name DeviceDriver was proposed to avoid that, but Tom et. al is 100% right that people want Device and that's good by me. To avoid the confusion and keep it separate from this re-work, we can just change the wording of our internal discussion of physical devices/simulators! I like backend personally (ty IBM) but we'll figure that out :p